# Metagenomics workshop - March 17th 2023

## Table of contents 

**Part 1. Basic set up of the computational environment**

1. Roar configuration
2. Accessing the Roar 
3. Submitting jobs to the Roar
4. Conda configuration

**Part 2. Metagenomics data processing**

1. Quality check and adapter removal
2. Taxonomy alignment (by MALT (MEGAN alignment tool))

---------------

## Part 1. Basic set up of the computational environment



### Roar configuration

Let's go over our [Wiki](https://github.com/DAWGPSU/DAWG_workshop/wiki/Roar-configuration) page to set up the Roar

***Check points***
```
1. Do you have your account requested and approved?
```

----------------

### Accessing the Roar (using interactive desktop)

Let's practice different ways to access to the Roar command line.

#### Using ssh

Please note that we are connecting the "**submission portal**," the batch portion of Roar. You CANNOT use the full resources of computation power if you run your job from the submission portal, and it is highly not recommended. In the later part of the instruction, we will also cover how to submit your job using the "qsub" command. You can still navigate your files, and delete and organize files. 

**For Mac and Linux**

1. Open 'terminal'
2. Type the following command
```
ssh <your penn state account(i.e., abc1234)>@submit.aci.ics.psu.edu
```
3. Once you run this command, the system will ask you for your Penn State account password 2-factor authentication via Duo app. Once authentication is completed, you should be connected to Roar. 

**NOTE: You WILL NOT see anything typed even if you are typing your password, don't worry and keep going and press "enter"**

**For Windows**

For Windows users, you need external programs to connect SSH ([Putty](https://www.putty.org)). Alternatively, you can install and use Ubuntu on your machine. If you have ubuntu installed from the app store, you can follow the instruction above as Linux OS.

1. Install and open the Putty

![image](https://user-images.githubusercontent.com/126822453/222592163-9d18ac84-518b-43f2-bb40-84b273ecc0ba.png)

2. You can log in to the Roar by using the below - 
  * Host Name - submit.aci.ics.psu.edu
  * Connection type - SSH 
  * Username (login as) - your PSU username (i.e., abc1234)
  * Password - your PSU password

3. The system will ask you for 2-factor authentication via the Duo app once you open the windows. Once authentication is completed, you should be connected to Roar. 

#### Using ACI shell access

You can connect to the submission portal via web-based shell access. It is very similar to SSH, so consider using this approach if you have a problem connecting via SSH. 

1. Go to https://portal2.aci.ics.psu.edu
2. Click "Clusters" from the top menu, then click ">_ACI Shell Access"
3. You will see a very similar screen than you connect via SSH. Once you open the windows, the system will ask you for 2-factor authentication via the Duo app. Once authentication is completed, you should be connected to Roar.

#### Using interactive desktop

Unlike SSH or shell access to submit.aci.ics.psu.edu, opening an interactive desktop means you are **[requesting resources](https://www.icds.psu.edu/computing-services/roar-user-guide/roar-system-layout/)** from the Roar. You can run your code, get the results, and monitor your job. It is a "virtual" computer you are borrowing from the university. 

1. Go to https://portal2.aci.ics.psu.edu
2. Click "Interactive Apps" from the top menu, then click "ACI RHEL7 Interactive Desktop."
3. You can launch an interactive desktop using the below configurations - 
  * Desktop Environment - MATE Gnome2
  * Allocation - Open (OR choose paid allocation if you have any)
  * Number of hours - You can use up to 48 hours if using Open allocation, and more time can be used if you use paid allocation.
  * Node type - ACI-i is only available for open account submissions. Other node types can be used too.

***Check point***
```
1. Did you launch an interactive desktop for 2 hours for this workshop?
```

----------------

### Submitting jobs to the Roar

Let's go again to our [Wiki page](https://github.com/DAWGPSU/DAWG_workshop/wiki/Submitting-Job-to-Roar) to go over multiple ways to submit jobs to the Roar

Today, we will use the interactive desktop to code in real time. 

----------------

### Conda configuration

![image](https://user-images.githubusercontent.com/126822453/222797077-f2a1cafb-9b79-4e0f-9396-77df31504515.png)

We need many bioinformatics programs to be installed to process metagenomics sequencing data. [Conda](https://docs.conda.io/en/latest/) is a package and environment manager to make installing, executing, and managing multiple packages easier. 

We are following instructions from [Conda website](https://conda.io/projects/conda/en/latest/user-guide/getting-started.html)

**Install miniconda3**

```
cd ~/scratch
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh 
sh Miniconda3-latest-Linux-x86_64.sh -b ## URL should be changed based on your OS 
source ~/.bashrc
```

Add [bioconda](https://bioconda.github.io) and [conda-forge](https://conda-forge.org) channel to the conda package manager

```
conda config --add channels bioconda
conda config --add channels conda-forge
```

**Set up conda environment with required packages - and additional programs installation with initial configuration**

A conda environment is where you can install a collection of conda packages. The main advantage of using an environment is to avoid the different versions of the dependencies from various bioinformatics tools because you only can use the installed packages in the specific environment. 

Create an environment

```
conda create -y --name DAWG python=3.8 # Create an environment named "DAWG" and install "python (version 3.8)"
conda activate DAWG # Activate your environment; now you only use packages installed in your environment
```
##  1-1. mamba installation

Even though you can download and install the bioinformatics package via the conda installer, I prefer to use Mamba, which is a reimplementation of the conda package manager in C++. It provides faster and more convenient installation 

You need conda to install mamba. As I mentioned, you can always use the conda installer to install conda packages. 

```
conda install -c conda-forge mamba
```

You can check if your mamba installation was successful by typing the below code

```
mamba -h
```
##  1-2. package installation

Now you can use the mamba installer (or conda if you prefer) to install bioinformatics packages. You can search and find if your bioinformatics tool was uploaded in conda repository (https://anaconda.org). 

1. Go to https://anaconda.org

![image](https://user-images.githubusercontent.com/126822453/222825235-44ee3599-2e6f-4392-8072-c6baf9ac58e5.png)

2. Search the package you want to install
3. Click the package name

4. follow the conda install instructions; please replace "conda" with "mamba" if you want to use the mamba installer

![image](https://user-images.githubusercontent.com/126822453/222825138-00e8aa3e-1ca4-43e7-8b43-0d57b517410e.png)

For example, if you want to install the "AdapterRemoval" package, you can type below -

```
## using conda installer
conda install -c bioconda adapterremoval

## using mamba install
mamba install -c bioconda adapterremoval
```

The results should be the same. One fantastic thing about conda/mamba is that they download and install all the required and optional dependencies together with the package, so you don't have to have a headache installing individual dependencies. 


5. Check your installation 
```
adapterremoval --help
```


***Check point***
```
1. Did you install conda and mamba successfully?
2. Did you install AdapterRemoval successfully?
```

------------------------------------

Now this is the end of the part1. Let's have a deep breath! 

![image](https://user-images.githubusercontent.com/126822453/224905728-39bab550-e933-42f0-876c-3150ab5b7cfa.png)

-------------------------------------

## Part 2. Basic set up of the computational environment

### Download data into your "scratch" folder

```
cd ~/scratch
wget --no-check-certificate "https://pennstateoffice365-my.sharepoint.com/:u:/g/personal/tuc289_psu_edu/EfPeT7R3E2VHmod0UcrNZlgBNKBHH-K5jvbIGY4PXorToA?e=OsM0L7&download=1" -O DAWG_031723.zip
unzip DAWG_031723.zip -d DAWG_031723
cd DAWG_031723
```

### Quality check and adapter removal

#### Quality check by FastQC

#### Removing adapters

### Taxonomy alignment by MALT (Megan Alginment tool)

### Other taxonomy alignment/classification tools

#### Kraken2 and Bracken

#### mOTU

#### MetaPhlAn






